src/clr-angular/utils/smart-popover/providers/smart-popover-events.service.ts

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ *
+ */
+
+import { ElementRef, Injectable, Renderer2, Inject, OnDestroy } from '@angular/core';
+import { ClrSmartPopoverToggleService } from './smart-popover-toggle.service';
+import { Observable, fromEvent, Subscription } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { DOCUMENT } from '@angular/common';
+
+// https://github.com/angular/angular/issues/20351#issuecomment-344009887
+/** @dynamic */
+@Injectable()
+export class ClrSmartPopoverEventsService implements OnDestroy {
+  public outsideClickClose = true;
+  public scrollToClose = true;
+  private documentClickListener: () => void;
+  public ignoredEvent: any;
+  private subscriptions: Subscription[] = [];
+
+  constructor(
+    private renderer: Renderer2,
+    private smartOpenService: ClrSmartPopoverToggleService,
+    @Inject(DOCUMENT) private document: HTMLDocument
+  ) {
+    this.subscriptions.push(
+      smartOpenService.openChange.subscribe(open => {
+        if (open) {
+          this.addEscapeListener();
+          this.addClickListener();
+          this.addScrollListener();
+        } else {
+          this.removeAllEventListeners();
+        }
+      }),
+      smartOpenService.getEventChange().subscribe(event => {
+        // Remember the event that was used to open the content
+        this.ignoredEvent = event;
+      })
+    );
+  }
+
+  private scrollSubscription: Subscription;
+  public addScrollListener() {
+    if (this.scrollToClose) {
+      this.documentScroller = fromEvent(this.document, 'scroll', { capture: true });
+      this.scrollSubscription = this.documentScroller
+        .pipe(filter(this.testForSmartPopoverContentContainer))
+        .subscribe(() => {
+          this.smartOpenService.open = false;
+        });
+    } else {
+      // I think this is where dynamic re-positioning will be added
+      // Instead of testing like we do in the close pipe below
+      // we need to switch positioning to use an observable and then
+      // debounce the scroll events to recalculate content position in a performant way
+      // For now, ignore scrolling events.
+      return;
+    }
+  }
+
+  public removeScrollListener() {
+    if (this.documentScroller) {
+      this.scrollSubscription.unsubscribe();
+      delete this.documentScroller;
+    }
+  }
+
+  private testForSmartPopoverContentContainer(event: Event): boolean {
+    // Filter for the documentScroller observable event targets
+    let target: Element = <Element>event.target;
+    while (target.classList && target.parentElement.localName !== 'body') {
```

Comment:
This fn is used as a filter on line #51 which is a subscription to _every_scroll event that happens on the document. 

When the scroll event happens inside _.clr-popover-content_ element we want to ignore them (and return false so that the observable above does not fire. When its not false scrolling occur outside of the _clr-popover-content_ element and we need to close the popover. In the future this behavior could be enhanced to move the content with the anchor if the consumer wants the content to stay open on scroll. 

This is a limitation with this approach. Before we got scrolling for free b/c the position calcuations were made for and anchor inside a relatively positioned container. Now we cannot use css to keep the content next to the anchor if the user scrolls.

Reply:
I updated this section and left comments to make it more understandable.